src/Microsoft.Diagnostics.Monitoring.EventPipe/Counters/EventCounterPipeline.cs

```diff
@@ -0,0 +1,46 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using Microsoft.Diagnostics.NETCore.Client;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.Diagnostics.Monitoring.EventPipe
+{
+    public class EventCounterPipeline : EventSourcePipeline<EventPipeCounterPipelineSettings>
+    {
+        private readonly IEnumerable<IMetricsLogger> _metricsLogger;
+        private readonly CounterFilter _filter;
+
+        public EventCounterPipeline(DiagnosticsClient client,
+            EventPipeCounterPipelineSettings settings,
+            IEnumerable<IMetricsLogger> metricsLogger) : base(client, settings)
```

Comment:
My general thinking was that inputs and outputs to pipelines should be treated separately. For example, most inputs can be expressed as serializable pod objects. The outputs vary greatly across pipelines and even within the event pipeline. Over time, if we create things like PipelineFactory, I thought it would be easier to treat these separately and not mix inputs and outputs. Separating the pipelines into different types was to make the api surface clearer, but they could certainly be folded back together.

Reply:
I think some of the inputs aren't serializable either, for example all the EventPipePipelines are taking a DiagnosticClient constructor argument. This may be bad naming on my part that settings implied the type should be serializable. Thinking on it a little more, maybe it would be more intuitive if I had proposed the reverse... keep the strongly typed pipeline types and eliminate the strongly typed settings. This has even fewer total types and no implication that the pipeline itself is serializable. A simple use case might look like:
```C#
EventPipeCountersPipeline p = new EventPipeCountersPipeline(client, logger, event_counter_list);
await p.RunAsync();
```
If it was problematic to gather all the configuration arguments together for a single constructor call the pipeline could also accumulate configuration via properties/methods (basically playing the role of the settings object without having a separate type)
````C#
EventPipeCountersPipeline p = new EventPipeCountersPipeline();
ConfigurePipeline(p);
p.DiagnosticClient = GetClientForPid(pid);
p.MetricLogger = logger;
await p.RunAsync();

void ConfigurePipeline(EventPipeCountersPipeline p)
{
    p.AddCounterGroup(...)
    p.AddCounterGroup(...)
}
````

A more extensible serialization friendly use-case that we could implement in the future might look like:
````C#
// at startup something needs to discover some factory implementations and register them
class EventCountersPipelineFactory : PipelineFactory
{
    override string PipelineTypeName => "Counters";
    override IPipeline CreatePipeline(IConfiguration configData, DiagnosticClient client, Stream outputStream)
    {
        MetricsLogger logger = new MetricsLogger(outputStream);
        return new EventPipeCountersPipeline(client, logger, ParseCounterList(configData));
    }
}

void RegisterFactory(PipelineFactory factory)
{
    m_factories[factory.PipelineTypeName] = factory;
}

// later a request arrives and we deserialize it and run it
RunPipeline(string configText, Stream outputStream)
{
    IConfiguration config = ParseConfig(configText);
    IPipelineFactory factory = m_factories[config["PipelineType"]];
    DiagnosticClient client = m_clients.GetClient(config["ClientInfo"]);
    IPipeline p = factory.CreatePipeline(config["PipelineData"], client, outputStream);
    await p.RunAsync();
}
````

Not sure if this sounds reasonable or dubious, but happy to chat if that will help shake this out.

Reply:
I think the factory approach is ideal. Your design of registering factories based on names is pretty much exactly where I would like to land in the end since I think it would work really well for our trigger scenario. But in that context, I was still keeping inputs and outputs separated from each other. (There's also the issue of semantic understanding of those outputs by whatever is creating the pipeline).